src/Build/Graph/ParallelWorkSet.cs

```diff
@@ -115,51 +118,35 @@ internal void AddWork(TKey key, Func<TResult> workFunc)
         }
 
         /// <summary>
-        /// Marks the work set as completed.
+        /// Assists processing items until all the items added to the queue are processed, completes the work set, and
+        /// propagates any exceptions thrown by workers.
         /// </summary>
-        internal void Complete()
+        internal void WaitForAllWorkAndComplete()
         {
-            bool schedulingCompleted = Volatile.Read(ref _isSchedulingCompleted);
-            if (schedulingCompleted)
+            if (IsCompleted)
             {
                 return;
             }
+            while (!_cancellationToken.IsCancellationRequested && Interlocked.Read(ref _pendingCount) > 0)
```

Comment:
If the last remaining work item is on another thread, won't this spin here really hard? Or is `_pendingCount` only for this thread?

Reply:
It will when there isn't another item in the queue (most likely start/end of the graph build), although it only effects the main/completion thread. I'm scoping this to another PR since it's a bit more involved to properly fix.

Reply:
Rather then spinning on a `_pendingCount`, would spinning on a `remainingUnassignedWorkCount` fix it?

Reply:
But then you'd still need to block if all remaining work has been assigned, to wait for it all to finish.

Reply:
```remainingUnassignedWorkCount``` is essentially ```_queue.Count```. Spinning on that breaks on the scenario where a worker dequeues the last work item in the queue but that work item recursively enqueues more work at the end, then the completion thread exits the loop too early.